### PR TITLE
Select/combobox built-in labels

### PR DESCRIPTION
**Pull request by msmithNI**
# Pull Request

## 🤨 Rationale

Implementation for #2183 . (Work item includes figuring out usage guidance, so won't close it with this PR yet)

## 👩‍💻 Implementation

See updated HLD in this PR.

## 🧪 Testing

- Local testing in Storybook
- Added autotests to cover the select's new (internal) labelContent property and that it sets aria-label (similar to those in `list-option-group.spec.ts`)
- Storybook: Added the 'height test' stories present in other controls that have labels
- Inspected Chromatic diffs. The changes are all expected:
   - Disabled controls now have disabled labels
   - If a `width` style is set on the control (as it is in the theme matrix tests), labels will wrap to multiple lines if needed. This matches existing control behavior (see text area / number field stories)
- Verified pre-existing accessibility violations for [select](https://nimble.ni.dev/storybook/?path=/story/components-select--select) / [combobox](https://nimble.ni.dev/storybook/?path=/story/components-combobox--underline-combobox) are resolved

## ✅ Checklist

- [x] I have updated the project documentation to reflect my changes or determined no changes are needed.
   - Updated Storybook docs to document label support for select/combobox

### CODE REVIEW

File: packages/nimble-components/src/select/specs/built-in-labels.md
Comment (msmithNI) [inline packages/nimble-components/src/select/specs/built-in-labels.md:42]: After further implementation / style tweaks, I found that I didn't need to add a new container div after all, so I simplified the implementation and HLD here. There's also no changes to the rich text mention listbox template, with the updated approach.

Initial issues I hit when prototyping were (1) to get the `div.control` to still size correctly, (2) focus/ focus-visible styling, including getting the underlines to still center, but those issues are fixed now.

File: packages/storybook/src/nimble/combobox/combobox.stories.ts
Comment (msmithNI) [inline packages/storybook/src/nimble/combobox/combobox.stories.ts:118]: Note: Switching from `width` to `min-width` in these stories so that the control will grow with a long label, instead of wrapping the label (matches more of our other control stories).

File: packages/nimble-components/src/combobox/index.ts
Comment (msmithNI) [inline packages/nimble-components/src/combobox/index.ts:884]: From the HLD, this resolves an accessibility violation.

>  -   Elements must only use supported ARIA attributes: `nimble-combobox` `nimble-toggle-button`: ARIA attribute is not allowed: `aria-expanded="false"`
>     -   Can be resolved via `aria-hidden="true"` on the dropdown toggle button, as it's not keyboard-focusable.

**Comment by msmithNI** [comment]
@atmgrifter00 can you buddy this PR?

File: packages/nimble-components/src/select/template.ts
Comment (atmgrifter00) [inline packages/nimble-components/src/select/template.ts:61]: I would have expected the `Combobox` to need this as well.

File: packages/nimble-components/src/select/index.ts
Comment (atmgrifter00) [inline packages/nimble-components/src/select/index.ts:207]: Why doesn't the `Combobox` need this same implementation and set of tests?

File: packages/nimble-components/src/select/template.ts
Comment (fredvisser) [inline packages/nimble-components/src/select/template.ts:61]: I don't think the select [requires](https://stackoverflow.com/a/56775032/13331204) an explicit `aria-label`, because assistive technology can already infer the label from the slotted content. If that's the case, we should remove the `aria-label`.

I.e. The label and aria-label are both displayed in VoiceOver 

Screenshot 2024-08-01 at 10 30 25 AM

Comment (msmithNI) [reply packages/nimble-components/src/select/template.ts:61]: Without the `aria-label` set on the select, the select still has an accessibility violation about not having an accessible label. I ran a build with that change, [this Storybook build without aria-label](https://60e89457a987cf003efc0a5b-zgzmmrogzr.chromatic.com/?path=/story/components-select--select) shows the violation.
As I noted above, the select and combobox declare their ARIA role in different places (also in the shadow root for the combobox, vs on the template/element itself for the combobox), which may be the reason for the different behavior.

Good catch that the current changes can result in duplicate narration though. Since `aria-label` is set, I probably need to set `aria-hidden` on the `label` element. (I see that in [list-option-group which sets aria-label](https://github.com/ni/nimble/blob/main/packages/nimble-components/src/list-option-group/template.ts#L19) too).

Comment (msmithNI) [reply packages/nimble-components/src/select/template.ts:61]: Added `aria-hidden` to the `label` element, since `aria-label` is used.

Comment (atmgrifter00) [reply packages/nimble-components/src/select/template.ts:61]: There's still a bit of logic present to produce the `labelContent` string. I presume we've concluded that the results produced by that logic are unneeded for the `Combobox`?

**Review by atmgrifter00 — APPROVED** [review]

File: packages/nimble-components/src/select/index.ts
Comment (msmithNI) [reply packages/nimble-components/src/select/index.ts:207]: As discussed in the other comment threads, combobox doesn't need `aria-label` set to resolve its label accessibility violation, so it doesn't need `labelContent` (only used to set `aria-label`).

File: packages/nimble-components/src/select/template.ts
Comment (msmithNI) [reply packages/nimble-components/src/select/template.ts:61]: The logic of `labelContent` is just used to compute the value for `aria-label`, whereas combobox will just use the label text directly for that.

In terms of the computed `labelContent` value vs. what the label shows visually, I checked that both handle plain text and  `span` elements. One minor difference I see is that `<span>A B C</span><span>D E F</span>` shows `A B CD E F` visually in the label (vs `labelContent` of `A B C D E F`). But the way the label is shown matches how other existing controls (number field/ text field/etc) show it, so I think it's OK.

File: change/@ni-nimble-components-18b7e8bd-7ece-49a0-a662-2a8424f6bfdd.json
Comment (jattasNI) [inline change/@ni-nimble-components-18b7e8bd-7ece-49a0-a662-2a8424f6bfdd.json:1]: Interesting that hovering over the label causes the control to enter the hover state and that clicking it behaves the same as clicking the control. This was slightly unexpected to me (I thought maybe I could select and copy the label text, but that's not possible). However it matches the behavior of labels on text and number field so I think it's the right call.

No action needed, just observing in case someone else sees this and finds it unexpected like I did.

https://github.com/user-attachments/assets/63e44712-1c99-44f1-9d22-c2013139e6b0

File: packages/nimble-components/src/select/index.ts
Comment (rajsite) [inline packages/nimble-components/src/select/index.ts:216]: We have this logic duplicated slightly differently a couple times. Made a tech debt to clean it up. Could help to add this example to the issue so it's found easier after merge: https://github.com/ni/nimble/issues/2322

File: packages/storybook/src/nimble/select/select.stories.ts
Comment (jattasNI) [inline packages/storybook/src/nimble/select/select.stories.ts:162]: Reminder to work with @fredvisser to figure out usage guidance on when clients should use label vs placeholder vs neither. Doesn't have to block this PR but it'd be good to have a decision before closing the feature.

File: packages/nimble-components/src/combobox/template.ts
Comment (rajsite) [reply packages/nimble-components/src/combobox/template.ts:37]: Same issue for select

**Review by jattasNI — COMMENTED** [review]
Just reviewed storybook docs and behavior, still need to look at the implementation.

File: packages/nimble-components/src/combobox/template.ts
Comment (rajsite) [reply packages/nimble-components/src/combobox/template.ts:37]: Looks like we aren't consistent across controls. Switch label is selectable but text area and text field labels aren't. Maybe create a bug to make consistent.

**Review by rajsite — APPROVED** [review]

File: packages/storybook/src/nimble/combobox/combobox.stories.ts
Comment (msmithNI) [reply packages/storybook/src/nimble/combobox/combobox.stories.ts:118]: The theme matrix stories have wrapping labels because they set a fixed `width` on the control. This matches the other stories for controls with labels (and I think a fixed width makes sense for the theme matrix so everything stays in a grid).

Comment (rajsite) [reply packages/storybook/src/nimble/combobox/combobox.stories.ts:118]: Think it's expected in matrix tests so the controls layout nicely. But for the doc stories it's allowed to grow

File: change/@ni-nimble-components-18b7e8bd-7ece-49a0-a662-2a8424f6bfdd.json
Comment (rajsite) [reply change/@ni-nimble-components-18b7e8bd-7ece-49a0-a662-2a8424f6bfdd.json:1]: Looks like in chrome and firefox a labelled select gets focused when the label is clicked (instead of open) but hover state on the label does apply to control and label text is selectable in control: https://jsbin.com/viveyitene/2/edit?html,js,output

Made a separate comment about selection of label, hover seems aligned, don't feel strongly about click opening vs just focusing

File: packages/nimble-components/src/combobox/template.ts
Comment (rajsite) [reply packages/nimble-components/src/combobox/template.ts:37]: Based on the other comment we should have the issue align our controls on mouse selection, hover interactions, and click activation.

File: change/@ni-nimble-components-18b7e8bd-7ece-49a0-a662-2a8424f6bfdd.json
Comment (rajsite) [inline change/@ni-nimble-components-18b7e8bd-7ece-49a0-a662-2a8424f6bfdd.json:1]: Looks like the Angular and Blazor example apps should be updated so the selects and comboboxes have labels. Feel free to do that in this PR or a follow-up

File: packages/nimble-components/src/combobox/template.ts
Comment (rajsite) [reply packages/nimble-components/src/combobox/template.ts:37]: That issue should also probably capture some docs on labelled control micro interactions to validate are consistent.

**Review by jattasNI — APPROVED** [review]

**Review by mollykreis — APPROVED** [review]

File: packages/nimble-components/src/combobox/index.ts
Comment (msmithNI) [reply packages/nimble-components/src/combobox/index.ts]: Removed aria props

File: change/@ni-nimble-components-18b7e8bd-7ece-49a0-a662-2a8424f6bfdd.json
Comment (msmithNI) [reply change/@ni-nimble-components-18b7e8bd-7ece-49a0-a662-2a8424f6bfdd.json:1]: Updated Angular/Blazor examples with select/combobox labels.

File: packages/nimble-components/src/combobox/template.ts
Comment (msmithNI) [reply packages/nimble-components/src/combobox/template.ts:37]: I attempted to capture all of this discussion / document the inconsistencies in #2331 .

I considered moving the `user-select: none` style, but left it as-is for this PR, since the current behavior matches more of the other labeled controls, and it's currently a shared dropdown style. We can revisit as part of that issue, as needed.

File: change/@ni-nimble-components-18b7e8bd-7ece-49a0-a662-2a8424f6bfdd.json
Comment (msmithNI) [reply change/@ni-nimble-components-18b7e8bd-7ece-49a0-a662-2a8424f6bfdd.json:1]: Documented the current inconsistencies in #2331

File: packages/nimble-components/src/select/index.ts
Comment (msmithNI) [reply packages/nimble-components/src/select/index.ts:216]: Done

**Pull request merged**
